traitsui/qt4/video_editor.py

```diff
@@ -0,0 +1,360 @@
+        self.control = QVideoWidget()
+        self.control.setSizePolicy(
+            QSizePolicy.Expanding, QSizePolicy.Expanding
+        )
+        self.control.setBackgroundRole(QPalette.Window)
+
+        self.media_player = QMediaPlayer(None, QMediaPlayer.VideoSurface)
+        self._set_video_url()
+        self.media_player.setVideoOutput(self.control)
+        self.media_player.setMuted(self.muted)
+        self._state_changed()
+        self._aspect_ratio_changed()
+        self._muted_changed()
+        self._volume_changed()
+        self._playback_rate_changed()
+
+        self._connect_signals()
+
+        self.set_tooltip()
+
+    def dispose(self):
+        self._disconnect_signals()
+
+    def update_editor(self):
+        """Update the editor when the object trait changes externally."""
+        self._set_video_url()
+
+    # ------------------------------------------------------------------------
+    # Private interface
+    # ------------------------------------------------------------------------
+
+    def _connect_signals(self):
+        if self.media_player is not None:
+            self.media_player.stateChanged.connect(self._state_changed_emitted)
+            self.media_player.positionChanged.connect(
+                self._position_changed_emitted)
+            self.media_player.durationChanged.connect(
+                self._duration_changed_emitted)
+            self.media_player.error.connect(self._error_emitted)
+            self.media_player.mediaStatusChanged.connect(
+                self._media_status_changed_emitted)
+            self.media_player.bufferStatusChanged.connect(
+                self._buffer_status_changed_emitted)
+
+    def _disconnect_signals(self):
+        if self.media_player is not None:
+            self.media_player.stateChanged.disconnect(
+                self._state_changed_emitted)
+            self.media_player.positionChanged.disconnect(
+                self._position_changed_emitted)
+            self.media_player.durationChanged.disconnect(
+                self._duration_changed_emitted)
+            self.media_player.error.disconnect(self._error_emitted)
+            self.media_player.mediaStatusChanged.disconnect(
+                self._media_status_changed_emitted)
+            self.media_player.bufferStatusChanged.disconnect(
+                self._buffer_status_changed_emitted)
+
+    def _set_video_url(self):
+        qurl = QUrl.fromUserInput(self.value)
+        if qurl.isValid():
+            self.media_content = QMediaContent(qurl)
+            self.control.updateGeometry()
+        else:
+            self.media_content = QMediaContent(None)
+            self.control.updateGeometry()
+
+    # Signal handlers -------------------------------------------------------
+
+    def _state_changed_emitted(self, state):
+        self.state = reversed_state_map[state]
+
+    def _position_changed_emitted(self, position):
+        self.position = position / 1000.0
```

Comment:
Regarding @jvkersch's comment about avoiding a call to `self.media_player.setPosition` in `_position_changed`, perhaps this should be `self.trait_setq(position=position / 1000)`?

The point is to keep the trait in sync, not set the position of the media player, right?